paddle/fluid/framework/ir/layer_norm_fuse_pass.cc

```diff
@@ -0,0 +1,231 @@
+  IR_NODE_LINK_TO(op_node, node_var);
+}
+
+}  // namespace
+
+void LayerNormFusePass::ApplyImpl(Graph* graph) const {
+  PADDLE_ENFORCE_NOT_NULL(graph,
+                          platform::errors::InvalidArgument(
+                              "The input graph of "
+                              "LayerNormFusePass should not be nullptr."));
+  FusePassBase::Init(scope_name_, graph);
+
+  auto* scope = param_scope();
+  PADDLE_ENFORCE_NOT_NULL(
+      scope, platform::errors::InvalidArgument("Scope cannot be nullptr."));
+
+  GraphPatternDetector gpd;
+  patterns::LayerNorm layer_norm_pattern(gpd.mutable_pattern(), scope_name_);
+  layer_norm_pattern();
+
+  int found_layer_norm_count = 0;
+  auto handler = [&](const GraphPatternDetector::subgraph_t& subgraph,
+                     Graph* g) {
+    VLOG(4) << "Fuse LayerNorm from subgraph.";
+    GET_IR_NODE_FROM_SUBGRAPH(x, x, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(x_mean, x_mean, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(x_mean_out, x_mean_out, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(x_sub_mean, x_sub_mean, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(x_sub_mean_out, x_sub_mean_out,
+                              layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(sqr_pow, sqr_pow, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(x_sub_mean_sqr, x_sub_mean_sqr,
+                              layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(x_sub_mean_sqr_out, x_sub_mean_sqr_out,
+                              layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(std_dev, std_dev, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(std_dev_out, std_dev_out, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(eps, eps, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(std_dev_eps, std_dev_eps, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(std_dev_eps_out, std_dev_eps_out,
+                              layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(std_dev_eps_sqrt, std_dev_eps_sqrt,
+                              layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(std_dev_eps_sqrt_out, std_dev_eps_sqrt_out,
+                              layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(division, division, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(division_out, division_out, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(gamma, gamma, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(scale, scale, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(scale_out, scale_out, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(beta, beta, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(shift, shift, layer_norm_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(shift_out, shift_out, layer_norm_pattern);
+
+    auto* eps_tensor = scope->FindVar(eps->Name())->GetMutable<LoDTensor>();
+
+    // ------------------ subgraph node's validation ---------------------------
+    PADDLE_ENFORCE_EQ(
```

Comment:
It's a general comment: in fuse passes, if any condition precludes fusion, we try not to throw exceptions. Instead, we just skip fusing with appropriate message printed to the output. Please see if you can follow this approach in the fuse.

Reply:
Good point!